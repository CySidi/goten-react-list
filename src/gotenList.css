@font-face {
  font-family: 'Glyphicons Halflings';
  src: url('//cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/fonts/glyphicons-halflings-regular.eot');
  src: url('//cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/fonts/glyphicons-halflings-regular.eot?#iefix') format('embedded-opentype'),
       url('//cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/fonts/glyphicons-halflings-regular.woff2') format('woff2'),
       url('//cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/fonts/glyphicons-halflings-regular.woff') format('woff'),
       url('//cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/fonts/glyphicons-halflings-regular.ttf') format('truetype'),
       url('//cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/fonts/glyphicons-halflings-regular.svg#glyphicons_halflingsregular') format('svg');
}

.actions {
    float: right;
}

.actionTitle {
    padding-left: 5px;
}

.title {
    padding-right: 5px;
}

.button-icon {
    border: 0px; 
    background-color: rgba(0, 0, 0, 0);
}

.glyphicon {
  position: relative;
  top: 1px;
  display: inline-block;
  font: normal normal 16px/1 'Glyphicons Halflings';
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  margin-right: 4px;
}

.glyphicon-pencil:before{content:"\270f";}
.glyphicon-trash:before{content:"\e020";}
.glyphicon-search:before{content:"\e003";}