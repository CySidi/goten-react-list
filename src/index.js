import React, { Component } from 'react'

import Table from 'react-bootstrap/lib/Table'
import Glyphicon from 'react-bootstrap/lib/Glyphicon'
import Button from 'react-bootstrap/lib/Button'

import './gotenList.css'


const bootstrapLink = <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u"
    crossOrigin="anonymous" />

const GotenListState = {
    items: []
}

export class GotenList extends Component {
    constructor(props){
        super(props)
        this.state = GotenListState
        this.keyItem = 0
    }

    removeItems() {
        this.setState({items:[]})
    }

    addItem(item, actions) {
        this.setState(prevState => ({
            items: [...prevState.items, {
                item: item,
                key: this.keyItem++,
                ...actions
            }]
        }))
    }

    addItemIteratively(item, numberOfItems=1, actions) {
        for (let index = 0; index < numberOfItems; index++) {
            this.addItem(item, actions)
        }
    }
    
    render(){
        return (
            <div>
                {bootstrapLink}
                <Table striped condensed hover>
                    <thead>
                        <tr>
                            <th>{this.props.title}</th>
                            <th>
                                <label className='actions'>
                                    {this.props.actionsTitle}
                                </label>
                            </th>
                        </tr>
                    </thead>
                    <tbody>
                        {this._getItems()}
                    </tbody>
                </Table>
            </div>
        )
    }
    
    _removeItem = async (item) => {
        if (this.props.onRemove)
            this.props.onRemove()
        else if (item.onRemove)
            item.onRemove()
        this.setState(prevState => ({
            items: [...prevState.items.filter(prevItem => prevItem.key !== item.key)]
        }))
    }

    _getOneAction(logic, name) {
        return (
            <th>
                { logic &&
                <Button
                    bsClass='button-icon'
                    onClick={logic}
                >
                    <Glyphicon glyph={name} />
                </Button>
                }
            </th>
        )
    }

    _getActions(item) {
        return (
            <table className='actions'>
                <tbody>
                    <tr>
                        {item.onSearch ?
                            this._getOneAction(item.onSearch, 'search')
                            : this._getOneAction(this.props.onSearch, 'search')}
                        {item.onEdit ?
                            this._getOneAction(item.onEdit, 'pencil')
                            : this._getOneAction(this.props.onEdit, 'pencil')}
                        {this._getOneAction(_ => this._removeItem(item), 'trash')}
                    </tr>
                </tbody>
            </table>
        )
    }

    _getItems() {
        return this.state.items.map( item => {
            return (
                <tr key={item.key}>
                    <th>
                        {item.item}
                    </th>
                    <th>
                        {this._getActions(item)}
                    </th>
                </tr>
            )
        })
    }
}
