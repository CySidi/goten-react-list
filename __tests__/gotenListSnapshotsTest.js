import React from 'react'
import renderer from 'react-test-renderer'

import { GotenList } from '../src'


describe('GotenList snapshots', () => {

    it('without title and actions title', () => {
        const tree = renderer.create(
            <GotenList />
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })

    it('without title', () => {
        const tree = renderer.create(
            <GotenList
                actionsTitle='Actions'
            />
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })

    it('without actions title', () => {
        const tree = renderer.create(
            <GotenList
                title='Title'
            />
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })

    it('with title and action', () => {
        const tree = renderer.create(
            <GotenList
                title='Title'
                actionsTitle='Actions'
            />
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })

    it('with onSearch', () => {
        const tree = renderer.create(
            <GotenList
                onSearch={_ => null}
            />
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })

    it('with onEdit', () => {
        const tree = renderer.create(
            <GotenList
                onSearch={_ => null}
            />
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })

    it('with onRemove', () => {
        const tree = renderer.create(
            <GotenList
                onRemove={_ => null}
            />
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })

    it('with onEdit, onRemove and onSearch', () => {
        const tree = renderer.create(
            <GotenList
                onSearch={_ => null}
                onEdit={_ => null}
                onRemove={_ => null}
            />
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })

    it('with all titles and actions', () => {
        const tree = renderer.create(
            <GotenList
                title='Title'
                actionsTitle='actionsTitle'
                onRemove={_ => null}
                onSearch={_ => null}
                onEdit={_ => null}
            />
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })
})