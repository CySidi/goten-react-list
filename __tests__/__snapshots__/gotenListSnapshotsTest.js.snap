// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GotenList snapshots with actions colors 1`] = `
<table
  className="table table-striped table-condensed table-hover"
  width="100%"
>
  <thead
    align="left"
    className="table-thead"
  >
    <tr>
      <th />
      <th
        align="right"
      />
    </tr>
  </thead>
  <tbody
    align="left"
  />
</table>
`;

exports[`GotenList snapshots with alignItems 1`] = `
<table
  className="table table-striped table-condensed table-hover"
  width="100%"
>
  <thead
    align="center"
    className="table-thead"
  >
    <tr>
      <th>
        Title
      </th>
      <th
        align="right"
      />
    </tr>
  </thead>
  <tbody
    align="center"
  />
</table>
`;

exports[`GotenList snapshots with all titles and actions 1`] = `
<table
  className="table table-striped table-condensed table-hover"
  width="100%"
>
  <thead
    align="left"
    className="table-thead"
  >
    <tr>
      <th>
        Title1
      </th>
      <th>
        Title2
      </th>
      <th>
        Title3
      </th>
      <th
        align="right"
      >
        actionsTitle
      </th>
    </tr>
  </thead>
  <tbody
    align="left"
  />
</table>
`;

exports[`GotenList snapshots with array title 1`] = `
<table
  className="table table-striped table-condensed table-hover"
  width="100%"
>
  <thead
    align="left"
    className="table-thead"
  >
    <tr>
      <th>
        Title1
      </th>
      <th>
        Title2
      </th>
      <th>
        Title3
      </th>
      <th
        align="right"
      />
    </tr>
  </thead>
  <tbody
    align="left"
  />
</table>
`;

exports[`GotenList snapshots with customButtons 1`] = `
<table
  className="table table-striped table-condensed table-hover"
  width="100%"
>
  <thead
    align="center"
    className="table-thead"
  >
    <tr>
      <th>
        Title
      </th>
      <th
        align="right"
      />
    </tr>
  </thead>
  <tbody
    align="center"
  >
    <tr>
      <td>
        <td>
          text from column1
        </td>
        <td>
          text from column2
        </td>
      </td>
      <td
        align="right"
      >
        <button
          onClick={[Function]}
        >
          MyButtonSearch
        </button>
        <button
          onClick={[Function]}
        >
          MyButtonEdit
        </button>
        <button
          onClick={[Function]}
        >
          MyButtonRemove
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`GotenList snapshots with onEdit 1`] = `
<table
  className="table table-striped table-condensed table-hover"
  width="100%"
>
  <thead
    align="left"
    className="table-thead"
  >
    <tr>
      <th />
      <th
        align="right"
      />
    </tr>
  </thead>
  <tbody
    align="left"
  />
</table>
`;

exports[`GotenList snapshots with onEdit, onRemove and onSearch 1`] = `
<table
  className="table table-striped table-condensed table-hover"
  width="100%"
>
  <thead
    align="left"
    className="table-thead"
  >
    <tr>
      <th />
      <th
        align="right"
      />
    </tr>
  </thead>
  <tbody
    align="left"
  />
</table>
`;

exports[`GotenList snapshots with onRemove 1`] = `
<table
  className="table table-striped table-condensed table-hover"
  width="100%"
>
  <thead
    align="left"
    className="table-thead"
  >
    <tr>
      <th />
      <th
        align="right"
      />
    </tr>
  </thead>
  <tbody
    align="left"
  />
</table>
`;

exports[`GotenList snapshots with onSearch 1`] = `
<table
  className="table table-striped table-condensed table-hover"
  width="100%"
>
  <thead
    align="left"
    className="table-thead"
  >
    <tr>
      <th />
      <th
        align="right"
      />
    </tr>
  </thead>
  <tbody
    align="left"
  />
</table>
`;

exports[`GotenList snapshots with title and action 1`] = `
<table
  className="table table-striped table-condensed table-hover"
  width="100%"
>
  <thead
    align="left"
    className="table-thead"
  >
    <tr>
      <th>
        Title
      </th>
      <th
        align="right"
      >
        Actions
      </th>
    </tr>
  </thead>
  <tbody
    align="left"
  />
</table>
`;

exports[`GotenList snapshots with width 1`] = `
<table
  className="table table-striped table-condensed table-hover"
  width="50%"
>
  <thead
    align="left"
    className="table-thead"
  >
    <tr>
      <th>
        Title
      </th>
      <th
        align="right"
      />
    </tr>
  </thead>
  <tbody
    align="left"
  />
</table>
`;

exports[`GotenList snapshots without actions title 1`] = `
<table
  className="table table-striped table-condensed table-hover"
  width="100%"
>
  <thead
    align="left"
    className="table-thead"
  >
    <tr>
      <th>
        Title
      </th>
      <th
        align="right"
      />
    </tr>
  </thead>
  <tbody
    align="left"
  />
</table>
`;

exports[`GotenList snapshots without title 1`] = `
<table
  className="table table-striped table-condensed table-hover"
  width="100%"
>
  <thead
    align="left"
    className="table-thead"
  >
    <tr>
      <th />
      <th
        align="right"
      >
        Actions
      </th>
    </tr>
  </thead>
  <tbody
    align="left"
  />
</table>
`;

exports[`GotenList snapshots without title and actions title 1`] = `
<table
  className="table table-striped table-condensed table-hover"
  width="100%"
>
  <thead
    align="left"
    className="table-thead"
  >
    <tr>
      <th />
      <th
        align="right"
      />
    </tr>
  </thead>
  <tbody
    align="left"
  />
</table>
`;
